Use default CBMC flags in RMC

adpaco-aws opened the pull request:
### Description of changes: 

This PR adds a couple of features to the RMC script:
  1. The requested feature to have a set of default CBMC flags used by RMC, which as of now consists of `--unwinding-assertions`.
  2. The scanner class which is used for matching text within CBMC's output an edit it. At present, there is a single scanner instance that looks for failed unwinding assertions in CBMC's output and prints a tip if it founds any.

I wrote the scanner class while also thinking about issue #189 , we would only need to write another instance that replaces text based on the failing assertion pattern.

### Resolved issues:

Resolves #156 

### Call-outs:

<!-- 
Address any potentially confusing code. Is there code added that needs to be cleaned up later? Is there code that is missing because it’s still in development? 
-->
This PR is quite large because the goal was to solve #156. We can split it into two smaller PRs if that will make reviewing easier.

### Testing:

* How is this change tested? Existing regression, plus a couple of new (positive) `cbmc` tests for testing `--no-default-flags` and another `expected` that checks the unwind failure recommendation is printed.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

vecchiot-aws commented on scripts/rmc.py:
```diff
@@ -31,6 +48,17 @@ def delete_file(filename):
 
+# Add a set of default CBMC arguments
+def add_default_cbmc_flags(args):
+    if not args.no_default_flags:
```
Is there a reason we block on `no_default_flags` both here and in `rmc`/`cargo-rmc`?

adpaco-aws replied:
It's a mistake, thanks for pointing this out. The condition is always checked before calling this function.

nchong-at-aws commented:
Looks good Adrian! A request: a way for the user to enumerate the default flags so they can get the information without reading source.

danielsn commented:
I think we'd be better off breaking the default flags into groups:
`unwinding-assertions`
`memory-safety`
`overflow-safety`
Then, we could have a the following flags
`no-memory-safety-checks`
`no-overflow-safety-checks`
`no-unwinding-checks`
and a 
`no-default-checks` which implies all of them

adpaco-aws commented:
> I think we'd be better off breaking the default flags into groups:
> `unwinding-assertions`
> `memory-safety`
> `overflow-safety`

I have changed the proposal to include the mentioned groups, but we have to discuss the specific flags for each group. In particular, what memory safety checks we want to enable?

> Looks good Adrian! A request: a way for the user to enumerate the default flags so they can get the information without reading source.

Thanks! I have added a flag `--print-cbmc-flags` that prints what CBMC flags will be used in an RMC invocation. Unfortunately, this does not work if the user does not specify an input file (as in a regular invocation) since the parser does not allow it. We could revisit this in the future.

danielsn commented on scripts/cargo-rmc:
```diff
@@ -31,6 +31,11 @@ def main():
+    parser.add_argument("--no-memory-safety-checks", action="store_true", help="Disable default memory safety checks")
+    parser.add_argument("--no-overflow-safety-checks", action="store_true", help="Disable default overflow checks")
+    parser.add_argument("--no-unwinding-checks", action="store_true", help="Disable default unwinding checks")
+    parser.add_argument("--print-cbmc-flags", action="store_true", help="Print CBMC flags associated to an RMC invocation")
```
Maybe make this a `--dry-run` flag?

danielsn commented:
We might also want to consider
- [ ]  `--div-by-zero-check          enable division by zero checks`
- [ ] ` --conversion-check           check whether values can be represented after type cast`
- [ ] ` --undefined-shift-check      check shift greater than bit-width`
- [ ] `-float-overflow-check       check floating-point for +/-Inf`
- [ ] `--nan-check                  check floating-point for NaN`
- [ ] `--enum-range-check           checks that all enum type expressions have values in the enum range`
- [ ] `--pointer-primitive-check    checks that all pointers in pointer primitives are valid or null`
- [ ] ` --malloc-may-fail            allow malloc calls to return a null pointer`
- [ ] ` --malloc-fail-null           set malloc failure mode to return null`

danielsn commented on src/test/firecracker/virtio-block-parse/main.rs:
```diff
@@ -1,13 +1,10 @@
-// --unwinding-assertions
-//
+
+// rmc-flags: --no-overflow-checks
```
Whenever we remove a check, lets have a comment explaining why, and maybe a tracking issue if its worth it

danielsn commented on src/test/expected/dry-run/expected:
```diff
@@ -0,0 +1,3 @@
+rmc-rustc -Z codegen-backend=gotoc -Z symbol-mangling-version=v0 --cfg=rmc
+symtab2gb
+cbmc --pointer-check --pointer-primitive-check --bounds-check --conversion-check --div-by-zero-check --float-overflow-check --nan-check --pointer-overflow-check --signed-overflow-check --undefined-shift-check --unsigned-overflow-check --unwinding-assertions --function main
```
newline

danielsn commented on src/test/cbmc/Whitespace/main.rs:
```diff
@@ -1,7 +1,8 @@
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
-// cbmc-flags: --unwind 2 --unwinding-assertions
+// rmc-flags: --no-overflow-checks
```
why?

danielsn commented on src/test/cbmc/DynTrait/boxed_closure.rs:
```diff
@@ -1,6 +1,8 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
+//rmc-flags: --no-memory-safety-checks
```
why?

danielsn commented on scripts/rmc:
```diff
@@ -41,6 +46,10 @@ def main():
+    
+    # Add some CBMC flags by default unless `--no-default-checks` is being used
+    if not args.no_default_checks:
+        rmc.add_default_cbmc_flags(args)
```
it wasn't clear to me from the name that this only adds some default flags (controlled by args)

adpaco-aws replied:
Not sure I follow. What changes to this are you suggesting?

danielsn replied:
naming it something like `add_selected_default_cbmc_flags`

adpaco-aws replied:
Done.

danielsn commented on scripts/rmc:
```diff
@@ -27,6 +27,11 @@ def main():
     parser.add_argument("--srcdir", default=".")
     parser.add_argument("--wkdir", default=".")
+    parser.add_argument("--no-default-checks", action="store_true", help="Disable all default checks")
+    parser.add_argument("--no-memory-safety-checks", action="store_true", help="Disable default memory safety checks")
```
Do we want a way to selectively add these back in?

adpaco-aws replied:
The user can always add them back in with `rmc --no-memory-safety-checks file.rs -- <cbmc-flags>`

danielsn commented:
For the future, the `scanner` is a big enough change it probably deserves its own PR. But for this one its OK

adpaco-aws commented:
Updated with explanations on each test that removes default checks. CI is failing due to `rustfmt` but I think @avanhatt has is taking care of it in #304 - In the meantime, let me know if we need other changes.

> For the future, the `scanner` is a big enough change it probably deserves its own PR. But for this one its OK

Yes, I agree.

danielsn commented on scripts/rmc.py:
```diff
@@ -5,12 +5,40 @@
 RMC_RUSTC_EXE = "rmc-rustc"
 EXIT_CODE_SUCCESS = 0
 
+MEMORY_SAFETY_CHECKS = ["--pointer-check",
+                        "--pointer-primitive-check",
+                        "--bounds-check"]
```
put these in alphabetical order

danielsn commented on scripts/rmc.py:
```diff
@@ -31,11 +59,29 @@ def delete_file(filename):
+    for arg in flags:
+        # This behavior must be reviewed if the set of flags is extended
+        if arg in args.cbmc_args:
+            print("WARNING: Default CBMC argument `{}` not added (already specified)".format(arg))
```
might be useful to also print the `cbmc_args` here

adpaco-aws replied:
Have to disagree with this. I expect the user to make a dry run if he really wants to see them. Plus, at this point `cbmc_args` is not complete (there may be other flags to add) so it could be misleading.

danielsn commented on scripts/rmc.py:
```diff
@@ -117,27 +186,27 @@ def run_visualize(cbmc_filename, cbmc_args, verbose=False, quiet=False, keep_tem
-        return run_cmd(cbmc_cmd, label="cbmc", output_to=output_to, verbose=verbose, quiet=quiet)
+        return run_cmd(cbmc_cmd, label="cbmc", output_to=output_to, verbose=verbose, quiet=quiet, dry_run=dry_run)
 
     cbmc_xml_args = cbmc_args + ["--xml-ui"]
```
putting all the property checks into the coverage checks is going to be super expensive. Lets split these out.

adpaco-aws replied:
Added another argument here for coverage checks even though I pass an empty list. The starter kit does not include any by default except for `--malloc-may-fail` and `--malloc-fail-null`, which we did not consider as default flags.
